src/client.rs

```diff
@@ -15,6 +15,37 @@ use super::http::{BaseClient, Query};
 use super::json_insert;
 use super::model::*;
 use super::oauth2::{Credentials, OAuth, Token};
+use itertools::Itertools;
+
+pub trait TryJoin: Itertools {
+    fn try_join<T, E>(&mut self, sep: &str) -> Result<String, E>
+    where
+        Self: Iterator<Item = Result<T, E>>,
+        T: AsRef<str>,
+    {
+        let (size, _) = self.size_hint();
+        let cap = size * sep.len();
+
+        self.fold_results(String::with_capacity(cap), |ids, id| {
+            ids + id.as_ref() + sep
+        })
+        .map(|mut ids| {
+            ids.pop();
+            ids
+        })
+    }
+
+    fn map_try_join<T, E, R, F>(&mut self, sep: &str, func: F) -> Result<String, E>
+    where
+        Self: Iterator<Item = T>,
+        F: Fn(T) -> Result<R, E>,
+        R: AsRef<str>,
+    {
+        self.map(func).try_join(sep)
+    }
+}
+
+impl<T> TryJoin for T where T: Itertools {}
```

Comment:
I am a little bit confused about the `TryJoin` trait and `may_try_join` function, what kind of problem will it solve ?

As for `itertools` crate, is it possible to replace it with a customized trait instead of introducing a new crate? Since we are working on reducing the compile time of `rspotify` :)

Reply:
It reduces code duplication, it just repeats all the way down in half of the API methods.

Reply:
As for `itertools` dependency, I'll try to remove it.

Reply:
I rewrote this trait into a single method.